File: app/src/main/java/com/cradle/neptune/view/ui/intro/IntroSectionsPagerAdapter.java
package com.cradle.neptune.view.ui.intro;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import androidx.core.app.Fragment;
import androidx.core.app.FragmentManager;
import androidx.core.app.FragmentPagerAdapter;

import com.cradle.neptune.R;
import com.cradle.neptune.utilitiles.Util;

/**
 * A [FragmentPagerAdapter] that returns a fragment corresponding to
 * one of the sections/tabs/pages.
 */
public class IntroSectionsPagerAdapter extends FragmentPagerAdapter {
    public static final int TAB_NUMBER_WELCOME = 0;
    public static final int TAB_NUMBER_PERMISSIONS = 1;
    public static final int TAB_NUMBER_PRIVACY_POLICY = 2;

    @StringRes
    private static final int[] TAB_TITLES = new int[]{
            R.string.intro_tab_welcome,
            R.string.intro_tab_permissions,
            R.string.intro_tab_privacy_policy
    };
    private final Context mContext;
    private Fragment[] fragments = new Fragment[TAB_TITLES.length];

    public IntroSectionsPagerAdapter(Context context, FragmentManager fm) {
        super(fm);
        mContext = context;
    }

    @Override
    public Fragment getItem(int position) {
        // cache created fragments so we can call their onMyBeingDisplayed() and onMyBeingHidden() methods.
        Fragment thisFragment = fragments[position];

        if (thisFragment == null) {
            // getItem is called to instantiate the fragment for the given page.
            // Return instance of required fragments
            switch (position) {
                case TAB_NUMBER_WELCOME:
                    thisFragment = WelcomeFragment.newInstance();
                    break;
                case TAB_NUMBER_PERMISSIONS:
                    thisFragment = PermissionsFragment.newInstance();
                    break;
                case TAB_NUMBER_PRIVACY_POLICY:
                    thisFragment = PrivacyPolicyFragment.newInstance();
                    break;
                default:
                    Util.ensure(false);
                    return null;
            }

            fragments[position] = thisFragment;
        }

        return thisFragment;
    }

    @Nullable
    @Override
    public CharSequence getPageTitle(int position) {
        return mContext.getResources().getString(TAB_TITLES[position]);
    }

    @Override
    public int getCount() {
        // Show # accessible pages.
        return TAB_TITLES.length;
    }
}

File: app/src/main/java/com/cradleplatform/neptune/adapters/introduction/IntroSectionsPagerAdapter.kt
package com.cradleplatform.neptune.adapters.introduction

import android.content.Context
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import com.cradleplatform.neptune.R
import com.cradleplatform.neptune.utilities.Util
import com.cradleplatform.neptune.fragments.introduction.PermissionsFragment
import com.cradleplatform.neptune.fragments.introduction.PrivacyPolicyFragment

/**
 * A [FragmentPagerAdapter] that returns a fragment corresponding to
 * one of the sections/tabs/pages.
 */
class IntroSectionsPagerAdapter(
    private val mContext: Context,
    fm: FragmentManager?
) : FragmentPagerAdapter(fm!!) {
    private val fragments =
        arrayOfNulls<Fragment>(tabTitles.size)

    override fun getItem(position: Int): Fragment {
        // cache created fragments so we can call their onMyBeingDisplayed() and onMyBeingHidden() methods.
        var thisFragment = fragments[position]
        if (thisFragment == null) {
            // getItem is called to instantiate the fragment for the given page.
            // Return instance of required fragments
            thisFragment = when (position) {
                TAB_NUMBER_PERMISSIONS -> PermissionsFragment.newInstance()
                TAB_NUMBER_PRIVACY_POLICY -> PrivacyPolicyFragment.newInstance()
                else -> {
                    Util.ensure(false)
                    PermissionsFragment.newInstance()
                }
            }
            fragments[position] = thisFragment
        }
        return thisFragment!!
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return mContext.resources
            .getString(tabTitles[position])
    }

    override fun getCount(): Int {
        // Show # accessible pages.
        return tabTitles.size
    }

    companion object {
        const val TAB_NUMBER_PERMISSIONS = 0
        const val TAB_NUMBER_PRIVACY_POLICY = 1

        @StringRes
        private val tabTitles = intArrayOf(
            R.string.intro_tab_permissions,
            R.string.intro_tab_privacy_policy
        )
    }
}
